### PR TITLE
feat: add basic rocksdb package

### PR DESCRIPTION
**Pull request by gocarlos**
Specify library name and version:  **rocksdb/6.4.6**

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by gocarlos** [comment]
TODO: according to the [INSTALL.md](https://github.com/facebook/rocksdb/blob/master/INSTALL.md#dependencies) compression is optional, should add this to the recipe

File: recipes/rocksdb/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
+from conans import ConanFile, CMake, tools
+
+
+class RocksDB(ConanFile):
+    name = "rocksdb"
+    homepage = "https://github.com/facebook/rocksdb"
+    license = "GPLv2"
```
Comment (uilianries) [inline recipes/rocksdb/all/conanfile.py]:
```suggestion
    license = "GPL-2.0"
```

File: recipes/rocksdb/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = "rocksdb-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.definitions["WITH_TESTS"] = False
+        cmake.definitions["WITH_GFLAGS"] = False
+        cmake.configure(source_folder=self._source_subfolder)
```
Comment (uilianries) [inline recipes/rocksdb/all/conanfile.py]:
```suggestion
        cmake.configure()
```

File: recipes/rocksdb/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        cmake.build()
+
+    def package(self):
+        self.copy(pattern="LICENSE*", dst="licenses", src=self._source_subfolder)
+        self.copy('*', dst='include',
+                  src='{}/include'.format(_source_subfolder))
+        self.copy("*rocksdb.lib", dst="lib", keep_path=False)
```
Comment (uilianries) [inline recipes/rocksdb/all/conanfile.py]: Why don't you use cmake.install() ?

**Comment by uilianries** [comment]
There are some requirements which can solved by Conan: https://github.com/facebook/rocksdb/blob/master/CMakeLists.txt#L54

**Comment by gocarlos** [comment]
> There are some requirements which can solved by Conan: https://github.com/facebook/rocksdb/blob/master/CMakeLists.txt#L54

done

File: recipes/rocksdb/all/conanfile.py (hunk 3 above)
Comment (gocarlos) [reply recipes/rocksdb/all/conanfile.py]: done

**Comment by gocarlos** [comment]
getting error 403 when trying to see the logs from jenkins....

File: recipes/rocksdb/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    def _configure_cmake(self):
+        cmake = CMake(self)
+        cmake.definitions["WITH_TESTS"] = False
+        cmake.definitions["WITH_GFLAGS"] = self.options.with_gflags
+        cmake.definitions["WITH_SNAPPY"] = self.options.with_snappy
+        cmake.definitions["WITH_LZ4"] = self.options.with_lz4
+        cmake.definitions["WITH_ZLIB"] = self.options.with_zlib
+        cmake.definitions["WITH_ZSTD"] = self.options.with_zstd
+        # not available yet in CCI
+        cmake.definitions["WITH_JEMALLOC"] = False
```
Comment (uilianries) [inline recipes/rocksdb/all/conanfile.py]: I've start an effort to package this one. It builds fine on Linux, but on Windows is a nightmare.

**Comment by gocarlos** [comment]
@uilianries I do not have access to a windows machine, so cannot really test the windows build. eventually you can test this :+1:

**Comment by uilianries** [comment]
@gocarlos I tested on Windows, it's falling to build build as shared lib. I'll revisit this week, it seems be an upstream bug

**Comment by gocarlos** [comment]
@uilianries any updates?

**Comment by uilianries** [comment]
@gocarlos https://github.com/gocarlos/conan-center-index/pull/2

**Comment by uilianries** [comment]
@gocarlos I think this new PR will fix all those errors: https://github.com/gocarlos/conan-center-index/pull/3

Happy holidays!

**Comment by uilianries** [comment]
And it failed again :disappointed:

**Comment by uilianries** [comment]
Interesting, I can't reproduce such error on Windows.

**Comment by danimtb** [comment]
Is this PR blocked by something else? @gocarlos

**Comment by gocarlos** [comment]
> R blocked by something

this works on my side, though I can only test on linux, as I do not have a windows system...
@uilianries was debugging the shared lib issues on windows...

I think we could also just prevent the shared lib issues on windows for now and merge this, should work on linux/mac/windows as static lib and on linux/mac as shared lib

**Comment by danimtb** [comment]
I am ok with that approach if handle it rains invalid configurations. At least having the recipe for linux and mac will more than nothing and anyone can work on the windows issue if needed

**Comment by gocarlos** [comment]
> I am ok with that approach if handle it rains invalid configurations. At least having the recipe for linux and mac will more than nothing and anyone can work on the windows issue if needed

can someone from the conan team do those smaller changes? is probably the fastest way to get it done. I guess the static lib under windows work well, its only the shared lib that makes issues

**Comment by danimtb** [comment]
@gocarlos cannot help you with those configurations right now. We can wait for @uilianries help or just set the invalid configuration in place for the moment. Sorry

File: recipes/rocksdb/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        if self.options.with_snappy:
+            self.requires("snappy/1.1.7")
+        if self.options.with_lz4:
+            self.requires("lz4/1.9.2")
+        if self.options.with_zlib:
+            self.requires("zlib/1.2.11")
+        if self.options.with_zstd:
+            self.requires("zstd/1.3.8")
+        if self.options.with_tbb:
+            self.requires("tbb/2019_u9")
```
Comment (gocarlos) [inline recipes/rocksdb/all/conanfile.py]: I think this recipe also depends on folly: e.g. https://github.com/conan-io/conan-center-index/pull/920

see https://github.com/facebook/rocksdb/tree/master/third-party

**Comment by gocarlos** [comment]
@uilianries the cmake file on rocksdb master is now fixed/better for shared libs etc.

**Comment by uilianries** [comment]
@gocarlos CI test failed due a sigabort:

```
rocksdb/20200307 (test package): Running test()
pure virtual method called
pure virtual method called
terminate called without an active exception
Aborted (core dumped)
```

**Comment by gocarlos** [comment]
> @gocarlos CI test failed due a sigabort:
> 
> ```
> rocksdb/20200307 (test package): Running test()
> pure virtual method called
> pure virtual method called
> terminate called without an active exception
> Aborted (core dumped)
> ```

yes, have no idea why this happens here

**Comment by gocarlos** [comment]
@uilianries can you have a look? On my side it works all

**Comment by ZaMaZaN4iK** [comment]
@uilianries any updates?

**Pull request closed without merging**